src/main/scala/org/graphframes/lib/ConnectedComponentsDF.scala

```diff
@@ -0,0 +1,465 @@
+  if (bigComponents.nonEmpty) {
+    println("Handling bigComponents...")
+    val bigComponentReps: Set[Long] = bigComponents.keys.toSet
+    val isBigRep = idxDataType match {
+      case _: LongType => udf { (vid: Long) => bigComponentReps.contains(vid) }
+      case _: IntegerType => udf { (vid: Int) => bigComponentReps.contains(vid.toLong) }
+    }
+    // bigComponentIds: representative vertex ID -> component ID
+    val bigComponentIds: Map[Long, Long] =
+      origVertices.where(isBigRep(col(ID))).select(ID, COMPONENT_ID).map {
+        case Row(id: Long, comp: Long) => id -> comp
+        case Row(id: Int, comp: Int) => id.toLong -> comp.toLong
+      }.collect().toMap
+    bigComponents.foreach { case (vid: Long, compVertices: DataFrame) =>
+      val componentId = bigComponentIds(vid)
+      val component = origVertices.join(compVertices, ID)
+        .select(col(ID), col(ORIG_ID),
+          lit(componentId).cast(idxDataType.simpleString).as(COMPONENT_ID))
+      v = v.unionAll(component)
+    }
+  }
+  // Handle vertices without edges.
+  v = origVertices.join(v, origVertices(ID) === v(ID), "left_outer")
+    .select(origVertices(ID).cast("long"), origVertices(ORIG_ID),
+      coalesce(v(COMPONENT_ID), origVertices(COMPONENT_ID)).cast("long").as(COMPONENT_ID))
+  // TODO: unpersist origVertices, bigComponents?
+  // Join COMPONENT_ID column with original vertices.
+  graph.vertices.join(v.select(col(ORIG_ID).as(ID), col(COMPONENT_ID)), ID)
+  }
+
+  /**
+   * Get the highest degree vertex in the graph.
+   *
+   * @param e  Edges in graph
+   * @param HIGH_DEGREE  Threshold for an acceptably high-degree vertex.  A vertex will be returned
+   *                     only if it has degree higher than this cutoff.
+   * @return Highest degree vertex, or None if no vertex has degree higher than HIGH_DEGREE.
+   */
+  def getHighestDegreeVertex(e: DataFrame, HIGH_DEGREE: Long): Option[Long] = {
+    val degrees: DataFrame = e.select(explode(array(SRC, DST)).as(ID))
+      .groupBy(ID).agg(count("*").cast("long").as("degree"))
+    val highDegrees: DataFrame = degrees.where(col("degree") > HIGH_DEGREE)
+      .groupBy(ID).agg(max("degree").as("degree"))
+    val vArray: Array[Long] = highDegrees.select(ID, "degree").take(1).map {
+      case Row(id: Long, degree: Long) => id
+      case Row(id: Int, degree: Long) => id.toLong
+    }
+    if (vArray.nonEmpty) Some(vArray(0)) else None
+  }
+
+  /**
+   * Handle extremely high-degree vertices in large graphs.
+   *
+   * TODO: Confirm that this works with IntegerType vertex indices.
+   *
+   * @param edges  Edges DataFrame.  This should ideally be cached already.
+   *               This should not include any self-edges (single-edge loops).
+   * @return (updated edges with columns "src" and "dst",
+   *          map from representative vertices of large components -> component vertices)
+   *         The edges and component DataFrames are cached.
+   */
+  def handleSkew(
```

Comment:
Note: I've been running tests without using this method.  There is a flag in run() which lets you skip this method.